### PR TITLE
Copy Akamai logs to S3 and import them into Redshift

### PR DESCRIPTION
**Pull request by kevinlitchfield**
Adds a service to copy all new logs from Akamai's FTP server to S3, and a service to import them from S3 into Redshift. They are both run with `script.rb`, which should be run via a Cron job once a day or however often.

Since logs can be large (over 100 MB) and are downloaded from FTP before being uploaded to S3, this needs a persistent filesystem.

### `LogFileCopier`

For each new log file:

- Download log to system running script
- Upload log to S3
- Remove local copy of log
- Move log file on FTP server to archive directory (`copied_to_s3`)

### `RedshiftImporter`

For each new log file:

- Import to Redshift
- Move log file on S3 to archive directory (`akamai_logs_imported`)

### CODE REVIEW

**Comment by ksheurs** [comment]
@kevinlitchfield diving into review now!

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+source 'https://rubygems.org'
+
+group :production do
+  gem 'dotenv'
+end
+
+group :production, :test do
+  gem 'aws-sdk'
```
Comment (ksheurs) [inline Gemfile]: i know AWS iterates on this a lot. maybe lock to the current version?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,38 @@
...
+- `REDSHIFT_USER`
+
+### Redshift database
+
+A table called `logs` needs to be manually created in the Redshift DB. To do this, you can connect to Redshift by running `psql -h <endpoint> -U <userid> -d <databasename> -p <port>`, then run:
+
+```sql
+CREATE TABLE logs(date VARCHAR(MAX), time VARCHAR(MAX), cs_ip VARCHAR(MAX), cs_method VARCHAR(MAX), cs_uri VARCHAR(MAX), sc_status VARCHAR(MAX), sc_bytes VARCHAR(MAX), time_taken VARCHAR(MAX), cs_referer VARCHAR(MAX), cs_user_agent VARCHAR(MAX), cs_cookie VARCHAR(MAX));
+```
+
```
Comment (ksheurs) [inline README.md]: Excellent README and documentation.

File: README.md [hunk 3]
```diff
@@ -1,3 +1,38 @@
...
+- `REDSHIFT_PASSWORD`
+- `REDSHIFT_PORT`
+- `REDSHIFT_USER`
+
+### Redshift database
+
+A table called `logs` needs to be manually created in the Redshift DB. To do this, you can connect to Redshift by running `psql -h <endpoint> -U <userid> -d <databasename> -p <port>`, then run:
+
+```sql
+CREATE TABLE logs(date VARCHAR(MAX), time VARCHAR(MAX), cs_ip VARCHAR(MAX), cs_method VARCHAR(MAX), cs_uri VARCHAR(MAX), sc_status VARCHAR(MAX), sc_bytes VARCHAR(MAX), time_taken VARCHAR(MAX), cs_referer VARCHAR(MAX), cs_user_agent VARCHAR(MAX), cs_cookie VARCHAR(MAX));
```
Comment (ksheurs) [inline README.md]: It's likely in the future we'll be using multiple CDN's. Should we make this table `akamai_logs`?

File: lib/ftp_client.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    def initialize
+      logger.log('Connecting to Akamai FTP server')
+      begin
+        @ftp_client = Net::FTP.new('vhxlogs.upload.akamai.com', 'vhx-logs', ENV['AKAMAI_FTP_PASSWORD'])
+        chdir_to_log_directory
+      rescue => exception
+        logger.error(exception, 'Error connecting to Akamai FTP server')
+      end
+      self
+    end
```
Comment (ksheurs) [inline lib/ftp_client.rb]: Not strong preference, but thoughts on the begin/rescue you did here vs the following format:

```ruby
def initialize
  # code
rescue => exception
  # log
end
```

File: lib/log_file_copier.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  def download_from_ftp(log_file)
+    logger.log("Downloading", indented: true)
+    ftp_client.getbinaryfile(log_file.path)
+  end
+
+  def copy_to_s3(log_file)
+    logger.log("Copying to S3", indented: true)
+    s3_client.object('akamai_logs/' + log_file.path).upload_file(log_file.path)
```
Comment (ksheurs) [inline lib/log_file_copier.rb]: are any of these files really big? worth considering http://docs.aws.amazon.com/AmazonS3/latest/dev/uploadobjusingmpu-ruby-sdk.html ?

File: lib/pg_client.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+class RedshiftImporter
+  class PgClient
+    def import_into_redshift(filename)
+      begin
+        pg.exec("COPY logs FROM 's3://vhxlogs/#{filename}' CREDENTIALS '#{redshift_credentials}' DELIMITER '\t' REGION 'us-east-1' FILLRECORD IGNOREHEADER 2 gzip;")
```
Comment (ksheurs) [inline lib/pg_client.rb]: in the past these are the options i used:

`DELIMITER '\t' MAXERROR 1 FILLRECORD REMOVEQUOTES TRUNCATECOLUMNS IGNOREHEADER 2 gzip`

these were all needed for various reasons from cloudfront (as opposed to akamai) log imports

File: Gemfile (hunk 1 above)
Comment (kevinlitchfield) [inline Gemfile]: Good point, will do.

File: README.md (hunk 3 above)
Comment (kevinlitchfield) [inline README.md]: Yes, for sure. `logs` was a bit of a placeholder, good to get that updated now.

File: lib/ftp_client.rb (hunk 4 above)
Comment (kevinlitchfield) [inline lib/ftp_client.rb]: Good point, that format is more consistent with what I've done elsewhere. I think I wrote it that way to isolate the begin/rescue to just the code that we'd expect to have a risk of error -- i.e., `logger.log` should never raise an exception, but there may be an FTP connection error that causes `Net::FTP` to raise an exception -- but there's no real benefit to that. Will change.

File: lib/log_file_copier.rb (hunk 5 above)
Comment (kevinlitchfield) [inline lib/log_file_copier.rb]: Looks like we are already doing this. `Aws::S3::Object#upload_file` does multipart uploads behind the scenes. See https://aws.amazon.com/blogs/developer/uploading-files-to-amazon-s3/

File: lib/pg_client.rb (hunk 6 above)
Comment (kevinlitchfield) [inline lib/pg_client.rb]: OK. Doesn't seem like `TRUNCATECOLUMNS` would hurt. Do you recall why `REMOVEQUOTES` was needed for Cloudfront?

File: lib/ftp_client.rb [hunk 7]
```diff
@@ -3,14 +3,14 @@
...
       begin
+        logger.log('Connecting to Akamai FTP server')
         @ftp_client = Net::FTP.new('vhxlogs.upload.akamai.com', 'vhx-logs', ENV['AKAMAI_FTP_PASSWORD'])
         chdir_to_log_directory
+        self
       rescue => exception
         logger.error(exception, 'Error connecting to Akamai FTP server')
       end
-      self
     end
```
Comment (ksheurs) [inline lib/ftp_client.rb]: with all these begin/rescue's i meant the following. again, not a big deal - just curious on thoughts:

```ruby
def initialize
  logger.log('Connecting to Akamai FTP server')
  @ftp_client = Net::FTP.new('vhxlogs.upload.akamai.com', 'vhx-logs', ENV['AKAMAI_FTP_PASSWORD'])
  chdir_to_log_directory
  self
rescue => exception
  logger.error(exception, 'Error connecting to Akamai FTP server')
end
```

File: lib/log_file_copier.rb (hunk 5 above)
Comment (ksheurs) [inline lib/log_file_copier.rb]: Oh that's great! That's cool it manages internally on whether it needs to use multi-part or not.

**Comment by ksheurs** [comment]
Left 2 more minor comments. I think this is good to go @kevinlitchfield.

Next steps could be working with Casey to get Looker setup to this for querying per site/channel. We should also close in on a decision in Asana re: the Akamai naming convention and get back to them. Anything I'm missing?

File: lib/ftp_client.rb (hunk 7 above)
Comment (kevinlitchfield) [inline lib/ftp_client.rb]: Yeah, I changed it to exactly that, I agree that makes sense. Should be exactly as you've written it in the latest diff.

**Pull request merged**